File: php/examples/zcsv_append.php
<?php
/**
 * Example of using ZippedCsv to append a csv file to an existing zipped
 * csv file. If the zipped csv file does not exist, it will be created.
 * 
 * Usage:
 *   php examples/zcsv_append.php <path_to_zipped_csv_file> <path_to_csv_file>
 */

include __DIR__ . '/../src/mc/csv.php';
include __DIR__ . '/../src/mc/zippedcsv.php';

use \mc\ZippedCsv;

if ($argc < 3) {
    echo "Usage: php examples/zcsv_append.php <path_to_zipped_csv_file> <path_to_csv_file>" . PHP_EOL;
    exit(1);
}

if(!file_exists($argv[2])) {
    echo "CSV file not found: " . $argv[2] . PHP_EOL;
    exit(1);
}

try {
    $zcsv = new ZippedCsv($argv[1]);
    $csv = new \mc\Csv();
    $csv->Load($argv[2]);
    $csvFileName = basename($argv[2]);
    $zcsv->AddCsv($csvFileName, $csv);
    $zcsv->Save();
    $zcsv->Close();
    echo "CSV file appended to zipped csv file: " . $argv[1] . PHP_EOL;
} catch (Exception $e) {
    echo "Error: " . $e->getMessage() . PHP_EOL;
    exit(1);
}

File: cpp/examples/zcsv_append.cpp
// zcsv_append — appends a row to a named table inside an existing .zcsv.
//
// Usage: zcsv_append <archive.zcsv> <table_name> <value1> [<value2> ...]

#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include "csv.h"
#include "zippedcsv.h"

int main(int argc, char* argv[]) {
  if (argc < 4) {
    std::cerr << "Usage: zcsv_append <archive.zcsv> <table_name> "
                 "<value1> [<value2> ...]\n";
    return EXIT_FAILURE;
  }

  std::string path(argv[1]);
  std::string table(argv[2]);

  std::vector<std::string> values;
  for (int i = 3; i < argc; ++i) values.emplace_back(argv[i]);

  mc::ZippedCsv z(path);

  mc::Csv* csv = nullptr;
  try {
    csv = &z.GetCsv(table);
  } catch (const std::invalid_argument& e) {
    std::cerr << "Error: " << e.what() << "\n";
    return EXIT_FAILURE;
  }

  int rc = csv->AddRow(values);
  if (rc != mc::Csv::kOk) {
    std::cerr << "Error adding row (code " << rc << "). Expected "
              << csv->TotalColumns() << " value(s), got "
              << values.size() << ".\n";
    return EXIT_FAILURE;
  }

  rc = z.Save();
  if (rc != 0) {
    std::cerr << "Error: could not save " << path << " (code " << rc << ")\n";
    return EXIT_FAILURE;
  }

  std::cout << "Appended row to '" << table << "' in " << path << ".\n";
  return EXIT_SUCCESS;
}
